### PR TITLE
Tree: fix deepFreeze for maps and sets

### PR DESCRIPTION
**Pull request by yann-achard-MS**
The test utility function `deepFreeze` was not freezing `Map` and `Set` objects as well as their contents.

### CODE REVIEW

File: packages/dds/tree/src/test/feature-libraries/deltaUtils.spec.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ const fooField = brand<FieldKey>("foo");
 const moveId = brandOpaque<Delta.MoveId>(42);
 
 function applyModifyToTree(node: MapTree, modify: Delta.Modify): Map<FieldKey, Delta.MarkList> {
-    deepFreeze(modify);
+    deepFreeze(modify, false);
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/feature-libraries/deltaUtils.spec.ts:27]: Why do we not want to freeze map methods?

File: packages/dds/tree/src/test/feature-libraries/sequence-field/compose.spec.ts [hunk 2]
```diff
@@ -17,17 +17,17 @@ const tag1: RevisionTag = brand(1);
 const tag2: RevisionTag = brand(2);
 
 function compose(changes: TestChangeset[]): TestChangeset {
-    changes.forEach(deepFreeze);
+    changes.forEach((c) => deepFreeze(c));
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/compose.spec.ts:20]: Is this change necessary?

File: packages/dds/tree/src/test/utils.ts [hunk 3]
```diff
@@ -38,14 +38,45 @@ import { brand, makeArray } from "../util";
...
-            deepFreeze(value);
+const frozenMethod = () => {
+    assert.fail("Object is frozen");
+};
+
+/**
+ * Recursively freezes the given object.
+ * @param object - The object to freeze.
+ * @param freezeMethods - Whether to freeze mutation methods on `Set` and `Map` instances.
+ * Passing `true` mutates the `Set` and `Map` instances encountered in the recursion.
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/utils.ts:49]: I think we give a little more detail on the contract of `deepFreeze` with respect to Map and Set depending on the value of `freezeMethods`, as well as briefly explaining why Map and Set are handled specially.

File: packages/dds/tree/src/test/utils.ts [hunk 4]
```diff
@@ -38,14 +38,45 @@ import { brand, makeArray } from "../util";
...
+ */
+export function deepFreeze<T>(object: T, freezeMethods: boolean = true): void {
+    if (object instanceof Map) {
+        for (const [key, value] of object.entries()) {
+            deepFreeze(key, freezeMethods);
+            deepFreeze(value, freezeMethods);
+        }
+        if (freezeMethods && !Object.isFrozen(object)) {
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/utils.ts:57]: If the map is already frozen, should we assert that its mutation methods are all equal to frozenMethod?

File: packages/dds/tree/src/test/utils.ts [hunk 5]
```diff
@@ -38,14 +38,45 @@ import { brand, makeArray } from "../util";
...
+            object.set = frozenMethod;
+            object.delete = frozenMethod;
+            object.clear = frozenMethod;
+        }
+    } else if (object instanceof Set) {
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/utils.ts:62]: Might not be worth doing, but considering factoring out the common logic for handling hidden values and methods for mutating them.

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/test/feature-libraries/sequence-field/compose.spec.ts (hunk 2 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/compose.spec.ts:20]: It is: `forEach` takes a delegate that takes 3 parameters, whose types do not match `deepFreeze`.

File: packages/dds/tree/src/test/utils.ts (hunk 3 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/test/utils.ts:49]: What is else is there to say abouf the contract of deepFreeze with respect to Map and Set depending on the value of freezeMethods?

Comment (alex-pardes) [reply packages/dds/tree/src/test/utils.ts:49]: Something like "Note: Calling `Object.freeze` on a Set or Map does not prevent it from being mutated. If `freezeMethods` is true, we remove the mutating methods from the Set or Map to ensure that its state cannot be changed. Regardless of the value of `freezeMethods`, `deepFreeze` will freeze the keys and values of a Map or Set."

I don't understand the motivation for the mode where we don't freeze the methods, so maybe the comment should hint at that.

**Review by alex-pardes — APPROVED** [review]
